Feature/cdutz/go ads ng (Streamlining of PLC4X API in PLC4Go and PLC4J)

chrisdutz opened the pull request:
Hi all ... feedback highly appreciated

chrisdutz committed:
chore(protocols): Added a pointer to a possibly interesting protocol specification.

- renamed the property "len" to "strLen" to avoid problems in Go

chrisdutz committed:
feat(ads): ADS Auto-Discovery

First working version that correctly sends out the requests to all valid IPs and then correctly processes the responses

chrisdutz committed:
feat(ads): ADS Auto-Discovery

- Switched to using a pre-allocated buffer

chrisdutz committed:
feat(ads): ADS Auto-Discovery

- Made the buffer allocation actually work nicely.

chrisdutz committed:
Merge remote-tracking branch 'origin/develop' into feature/cdutz/go-ads-ng

# Conflicts:
#	plc4go/internal/ads/Field.go
#	plc4go/internal/ads/Reader.go
#	plc4go/internal/ads/Writer.go

chrisdutz committed:
feat(protocols): Added a pointer to a possibly interesting protocol specification.

- Got the parsing of data-type and symbol-table working
- Tweaked the way the endianess is handles.

chrisdutz committed:
refactor(api) PLC4J API refactoring

- PlcBrowseItem's getChildren is now a Map
- PlcBrowseRequest now has a executeWithInterceptor method
- PlcBrowseResponse now supports multiple queries and therefore has:
    - a getQueryNames() method
    - the getResponseCode method was replaced with a getResponseCode(queryName) method
    - the getValues() method was replaced with a getValues(queryName) method
- PlcField:
    - Removed the getDefaultJavaType() as PlcValueType provides this
    - replaced String getPlcDataType() with PlcValueType getPlcValueType()
    - replaced int getNumberOfElements() with List<ArrayInfo> getArrayInfo()
- A new type PlcQuery for being the base of Browse queries
- PlcValueType now provides a default java type for each constant

Protocols:
- Extended the data-types of the following protocols to allow mapping to the PlcValueTypes of the API module:
    - canopen
    - genericcan

Adjusted the rest of the codebase accordingly.

chrisdutz committed:
Merge remote-tracking branch 'origin/develop' into feature/cdutz/go-ads-ng

chrisdutz committed:
refactor(api) PLC4Go API refactoring

- Continued porting the PLC4Go API to be on par with the updated PLC4J API
- Started porting most of the existing code, but espeically the Browse parts still need quite some work.

chrisdutz committed:
refactor(api) PLC4Go API refactoring

- Finished implementing the multi-level-array-info-parsing

chrisdutz committed:
refactor(api) PLC4Go API refactoring

- Continued cleaning up the code after the changes.

chrisdutz committed:
refactor(api) PLC4Go API refactoring

- Continued cleaning up the code after the changes.

chrisdutz committed:
refactor(api) PLC4Go API refactoring

- Continued cleaning up the code after the changes.

chrisdutz committed:
refactor(api) PLC4Go API refactoring

- Finished cleaning the compilation up

next will be getting the tests back green

chrisdutz committed:
refactor(api) PLC4Go API refactoring

- Finished fixing the go tests.

chrisdutz committed:
fix(test) Not all tests are executed when enabling the "jenkins-build" or the "with-all-tests" profiles.

- Got the "all-tests" profile working again.

sruehl commented on plc4go/internal/ads/Browser.go:
```diff
@@ -0,0 +1,270 @@
+}
+
+/*
+func LALALA(){
```
LALALA?

sruehl commented on plc4go/internal/ads/Browser.go:
```diff
@@ -0,0 +1,270 @@
+	if symbol, ok := m.symbolTable[symbolName]; !ok {
+		// Couldn't find the base symbol
+		return nil
+	} else if len(remainingSegments) == 0 {
```
else is unnecessary here as the if above has a return

sruehl commented on plc4go/internal/ads/Browser.go:
```diff
@@ -0,0 +1,270 @@
+	} else if len(remainingSegments) == 0 {
+		// TODO: Convert the symbol itself into a PlcBrowseField
+		return nil
+	} else {
```
else is unnecessary here as the if above has a return

sruehl commented on plc4go/internal/ads/Browser.go:
```diff
@@ -0,0 +1,270 @@
+		if symbolDataType, ok := m.dataTypeTable[symbolDataTypeName]; !ok {
+			// Couldn't find data type
+			return nil
+		} else {
```
else is unnecessary here as the if above has a return

sruehl commented on plc4go/internal/ads/Browser.go:
```diff
@@ -0,0 +1,270 @@
+			}
+		}
+	}
+	// TODO: Couldn't find property with the given name.
```
TODO: is that meant to be fixed?

sruehl commented on plc4go/internal/ads/Connection.go:
```diff
@@ -38,9 +42,53 @@ type Connection struct {
 	tracer             *spi.Tracer
 }
 
+/*
```
code outcommended here

sruehl commented on plc4go/internal/ads/Discoverer.go:
```diff
@@ -0,0 +1,250 @@
+				continue
+			}
+
+			if (discoveryResponse.GetRequestId() == 0) &&
```
early exit would improve readability

sruehl commented on plc4go/internal/ads/Discoverer.go:
```diff
@@ -0,0 +1,250 @@
+			// If we found an IPv4 address and this is not a loopback address,
+			// add it to the list of devices we will open ports and send discovery
+			// messages from.
+			if ipv4Addr != nil && !ipv4Addr.IsLoopback() {
```
early exit would improve readability

sruehl commented on plc4go/internal/ads/Discoverer.go:
```diff
@@ -0,0 +1,250 @@
+					}
+				}
+
+				if hostNameBlock != nil {
```
early exit would improve readability

sruehl commented on plc4go/internal/ads/Discoverer.go:
```diff
@@ -0,0 +1,250 @@
+		// Iterate over all addresses the current interface has configured
+		// For ADS we're only interested in IPv4 addresses, as it doesn't
+		// seem to work with IPv6.
+		for _, addr := range addrs {
```
In C-Bus discovery I moved those inner loops in functions, maybe that would help a bit here

sruehl commented on plc4go/internal/ads/Discoverer_test.go:
```diff
@@ -0,0 +1,36 @@
+)
+
+func TestDiscovererManual(t *testing.T) {
+	discoverer := NewDiscoverer()
```
skip() should be added here

sruehl commented on plc4go/internal/ads/DiscoveryMessageCodec.go:
```diff
@@ -0,0 +1,97 @@
+		log.Warn().Err(err).Msg("Got error reading")
+		return nil, nil
+	}
+	// TODO: maybe we return here a not enough error error
```
can this TODO be fixed?

sruehl commented on plc4go/internal/ads/Interactions.go:
```diff
@@ -0,0 +1,261 @@
+				amsTcpPacket := message.(model.AmsTCPPacket)
+				response := amsTcpPacket.GetUserdata().(model.AdsReadDeviceInfoResponse)
+				responseChannel <- response
+				close(responseChannel)
```
are you sure that is the right thing to do? closing the channel directly?

sruehl commented on plc4go/internal/ads/Interactions.go:
```diff
@@ -0,0 +1,261 @@
+			time.Second*5); err != nil {
+			close(responseChannel)
+		} else {
+			//			close(responseChannel)
```
outcommended code here

sruehl commented on plc4go/internal/ads/Interactions.go:
```diff
@@ -0,0 +1,261 @@
+			//			close(responseChannel)
+		}
+	}()
+	response, err := ReadWithTimeout(responseChannel)
```
shouldn't read with timeout respect the ctx too?

sruehl commented on plc4go/internal/ads/Interactions.go:
```diff
@@ -0,0 +1,261 @@
+	return response, nil
+}
+
+func ReadWithTimeout[T spi.Message](ch <-chan T) (T, error) {
```
this should get the ctx passed in an create a new ctx from that with timeout. Other than that time.After is leaking timers.

sruehl commented on plc4go/internal/ads/MessageTemplates.go:
```diff
@@ -0,0 +1,71 @@
+func (m *Connection) NewAdsReadDeviceInfoRequest() adsModel.AmsTCPPacket {
+	return adsModel.NewAmsTCPPacket(
+		adsModel.NewAdsReadDeviceInfoRequest(m.configuration.targetAmsNetId, uint16(adsModel.DefaultAmsPorts_RUNTIME_SYSTEM_01),
+			m.configuration.sourceAmsNetId, 800, 0, m.getInvokeId()))
```
what is the magic number 800 here?

sruehl commented on plc4go/internal/ads/Writer.go:
```diff
@@ -56,133 +55,135 @@ func NewWriter(messageCodec spi.MessageCodec, targetAmsNetId readWriteModel.AmsN
-		if needsResolving(field) {
-			adsField, err := castToSymbolicPlcFieldFromPlcField(field)
-			if err != nil {
+	/*	// TODO: handle context
```
outcommented code here

sruehl commented on plc4go/internal/cbus/Field.go:
```diff
@@ -44,9 +47,16 @@ func (s StatusRequestType) String() string {
 	return ""
 }
 
+type CbusField interface {
```
I would just name that Field here...

sruehl commented on plc4go/spi/default/DefaultBrowser.go:
```diff
@@ -21,22 +21,25 @@ package _default
 // DefaultBrowserRequirements adds required methods to Browser that are needed when using DefaultBrowser
 type DefaultBrowserRequirements interface {
-	BrowseField(ctx context.Context, browseRequest apiModel.PlcBrowseRequest, interceptor func(result apiModel.PlcBrowseEvent) bool, fieldName string, field apiModel.PlcField) (apiModel.PlcResponseCode, []apiModel.PlcBrowseFoundField)
+	InternalBrowse(ctx context.Context, browseRequest apiModel.PlcBrowseRequest, interceptor func(result apiModel.PlcBrowseItem) bool, queryName string, query apiModel.PlcQuery) (apiModel.PlcResponseCode, []apiModel.PlcBrowseItem)
```
what was the reason to prefix that internal?

sruehl commented on plc4go/spi/model/render_test.go:
```diff
@@ -44,11 +45,11 @@ func TestNonPanickingStrings(t *testing.T) {
 		&DefaultPlcBrowseRequestResult{},
 		&DefaultPlcBrowseRequest{},
-		&DefaultPlcBrowseQueryResult{},
+		//&DefaultPlcBrowseQueryResult{},
```
out commented code

sruehl commented on plc4go/spi/values/value_combination_test.go:
```diff
@@ -127,7 +127,7 @@ func TestCombinations(t *testing.T) {
 		},
 		{
-			name: apiValues.LIST,
+			name: apiValues.List,
```
Why is the constant suddenly lower case?

chrisdutz committed:
fix(test) Not all tests are executed when enabling the "jenkins-build" or the "with-all-tests" profiles.

- Got the RandomPackagesTest working again
- Currently commented out the "0-termination" in the ReadBufferByteBased
- Got the libpcapNg detection working on aarch64 mac

chrisdutz committed:
fix(test) Fix RandomPackagesTest

- Changed the ReadBufferByteBased to react on a system property "disable-string-0-termination" to make it work with the 0-terminating byte-buffer.

chrisdutz committed:
fix(test) Fix RandomPackagesTest

- Changed the ReadBufferByteBased to react on a system property "disable-string-0-termination" to make it work with the 0-terminating byte-buffer.

chrisdutz committed:
fix(test) Fix error in WriteBufferByteBased, which output the padding bytes before the string

-

chrisdutz committed:
fix(test) Fix error in WriteBufferByteBased, which output the padding bytes before the string

- Removed the disabling of 0-termination in the RandomPackageTest
- Added some comments on the reading and writing of strings (expected format)

chrisdutz committed:
Merge remote-tracking branch 'origin/develop' into feature/cdutz/go-ads-ng

# Conflicts:
#	plc4go/protocols/bacnetip/readwrite/model/NLMDisconnectConnectionToNetwork.go
#	plc4go/protocols/bacnetip/readwrite/model/NLMEstablishConnectionToNetwork.go
#	plc4go/protocols/bacnetip/readwrite/model/NLMIAmRouterToNetwork.go
#	plc4go/protocols/bacnetip/readwrite/model/NLMICouldBeRouterToNetwork.go
#	plc4go/protocols/bacnetip/readwrite/model/NLMInitalizeRoutingTable.go
#	plc4go/protocols/bacnetip/readwrite/model/NLMInitalizeRoutingTableAck.go
#	plc4go/protocols/bacnetip/readwrite/model/NLMRejectRouterToNetwork.go
#	plc4go/protocols/bacnetip/readwrite/model/NLMRouterAvailableToNetwork.go
#	plc4go/protocols/bacnetip/readwrite/model/NLMRouterBusyToNetwork.go
#	plc4go/protocols/bacnetip/readwrite/model/NLMWhoIsRouterToNetwork.go

chrisdutz committed:
refactor(api) PLC4J API refactoring

- Updated the builders to use the new addField and addFieldAddress naming

chrisdutz committed:
refactor(api) PLC4J API refactoring

- Applied Sebasitan's feedback

chrisdutz committed:
refactor(api) PLC4J API refactoring

- Applied Sebasitan's feedback

chrisdutz committed:
refactor(api) PLC4J API refactoring

- Applied Sebasitan's feedback

chrisdutz committed:
refactor(api) PLC4J API refactoring

- Renamed Field -> Tag in PLC4J

chrisdutz committed:
refactor(api) PLC4J API refactoring

- Renamed Field -> Tag in PLC4J

chrisdutz committed:
refactor(api) PLC4J API refactoring

- Renamed Field -> Tag in PLC4C

chrisdutz committed:
refactor(api) PLC4Go API refactoring

- Updated generated code for PLC4Go

chrisdutz committed:
refactor(api) PLC4J API refactoring

- Removed getPlcValueType and getArrayInfo from the PlcBrowseItem as this information is in the Tag
- Removed the PlcBrowseItemArrayInfo type, as this is replaced by ArrayInfo

chrisdutz committed:
refactor(api) PLC4Go API refactoring

- Continued the refactoring with PLC4Go

chrisdutz committed:
refactor(api) PLC4Go API refactoring

- Continued the refactoring with PLC4Go

chrisdutz committed:
Merge remote-tracking branch 'origin/develop' into feature/cdutz/go-ads-ng
